Explain "free" inputs and outputs in pipeline docs (#2824)
kedro-org/kedro: docs/source/nodes_and_pipelines/modular_pipelines.md

Reusing pipelines for slightly different purposes can be a real accelerator for teams and organisations when they reach a certain scale. In the real world, one could imagine pipelines with responsibilities like profiling or feature engineering being reused within the same project or even across projects via [micro-packaging](micro_packaging.md).

* In an ideal world, we would like to use the `cook_pipeline` twice as you would `defrost` and `grill` multiple meals beyond the `veg` currently hard-coded.
* Namespaces allow you to instantiate the same pipeline multiple times and keep operations isolated.
* Like one provides arguments to a class' constructor, you can provide overriding inputs/outputs/parameters to the `pipeline()` wrapper.

```{note}

> The set of overriding inputs and outputs must be a subset of the reused pipeline's "free" inputs and outputs, respectively. A free input is an input that isn't generated by a node in the pipeline, while a free output is an output that isn't consumed by a node in the pipeline. [`Pipeline.inputs()`](/kedro.pipeline.Pipeline.inputs) and [`Pipeline.outputs()`](/kedro.pipeline.Pipeline.outputs) can be used to list a pipeline's free inputs and outputs, respectively.

Author: ```suggestion
The set of overriding inputs and outputs must be a subset of the reused pipeline's "free" inputs and outputs, respectively. A free input is an input that isn't generated by a node in the pipeline, while a free output is an output that isn't consumed by a node in the pipeline. [`Pipeline.inputs()`](/kedro.pipeline.Pipeline.html#kedro.pipeline.Pipeline.inputs) and [`Pipeline.outputs()`](/kedro.pipeline.Pipeline.html#kedro.pipeline.Pipeline.outputs) can be used to list a pipeline's free inputs and outputs, respectively.
```

> The set of overriding inputs and outputs must be a subset of the reused pipeline's "free" inputs and outputs, respectively. A free input is an input that isn't generated by a node in the pipeline, while a free output is an output that isn't consumed by a node in the pipeline. {py:meth}`kedro.pipeline.Pipeline.inputs <Pipeline.inputs()>` and {py:meth}`kedro.pipeline.Pipeline.outputs <Pipeline.outputs()>` can be used to list a pipeline's free inputs and outputs, respectively.

Author: ```suggestion
The set of overriding inputs and outputs must be a subset of the reused pipeline's "free" inputs and outputs, respectively. A free input is an input that isn't generated by a node in the pipeline, while a free output is an output that isn't consumed by a node in the pipeline. {py:meth}`Pipeline.inputs() <kedro.pipeline.Pipeline.inputs>` and {py:meth}`Pipeline.outputs() <kedro.pipeline.Pipeline.outputs>` can be used to list a pipeline's free inputs and outputs, respectively.
```

> The set of overriding inputs and outputs must be a subset of the reused pipeline's "free" inputs and outputs, respectively. A free input is an input that isn't generated by a node in the pipeline, while a free output is an output that isn't consumed by a node in the pipeline. {py:meth}`Pipeline.inputs() <kedro.pipeline.Pipeline.inputs>` and {py:meth}`Pipeline.outputs() <kedro.pipeline.Pipeline.outputs>` can be used to list a pipeline's free inputs and outputs, respectively.

Reviewer: Yesssss ❤️

Reviewer: xref https://myst-parser.readthedocs.io/en/latest/syntax/cross-referencing.html#reference-roles in case you haven't seen this syntax before @stichbury

> The set of overriding inputs and outputs must be a subset of the reused pipeline's "free" inputs and outputs, respectively. A free input is an input that isn't generated by a node in the pipeline, while a free output is an output that isn't consumed by a node in the pipeline. [`Pipeline.inputs()`](/kedro.pipeline.Pipeline.html#kedro.pipeline.Pipeline.inputs) and [`Pipeline.outputs()`](/kedro.pipeline.Pipeline.html#kedro.pipeline.Pipeline.outputs) can be used to list a pipeline's free inputs and outputs, respectively.

Reviewer: ```suggestion
The set of overriding inputs and outputs must be a subset of the reused pipeline's "free" inputs and outputs, respectively. A free input is an input that isn't generated by a node in the pipeline, while a free output is an output that isn't consumed by a node in the pipeline. {py:meth}`kedro.pipeline.Pipeline.inputs <Pipeline.inputs()>` and {py:meth}`kedro.pipeline.Pipeline.outputs <Pipeline.outputs()>` can be used to list a pipeline's free inputs and outputs, respectively.
```

see https://myst-parser.readthedocs.io/en/latest/syntax/cross-referencing.html#reference-roles

Reviewer: (Didn't test this syntax locally but we can adjust, I _think_ it should work - we should not rely on direct URLs to the extent that we keep using Sphinx and MyST)

Author: Let me try!

Author: @astrojuanlu Seems to work with a minor modifications (switch the link and text in the syntax)!
